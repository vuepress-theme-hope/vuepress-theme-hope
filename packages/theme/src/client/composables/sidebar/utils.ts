import { h } from "vue";
import { useRoute } from "vue-router";
import { useIconPrefix } from "../themeData";
import AutoLink from "../../components/AutoLink";
import { isActiveLink } from "../../utils";

import type { VNode } from "vue";
import type {
  ResolvedSidebarItem,
  ResolvedSidebarHeaderItem,
  ResolvedSidebarGroupItem,
} from "../../../shared";

export const renderIcon = (icon?: string): VNode | null =>
  icon
    ? h("i", {
        class: ["iconfont", `${useIconPrefix().value}${icon}`],
      })
    : null;

export const renderItem = (
  config: ResolvedSidebarItem,
  props: VNode["props"]
): VNode => {
  return config.link
    ? // if the item has link, render it as `<AutoLink>`
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      h(AutoLink, {
        ...props,
        config,
      })
    : // if the item only has text, render it as `<p>`
      h("p", props, [renderIcon(config.icon), config.text]);
};

export const renderGroupHeader = (
  { collapsable, icon, text }: ResolvedSidebarGroupItem,
  open: boolean
): (VNode | null)[] => [
  // icon
  renderIcon(icon),
  // title
  h("span", { class: "title" }, text),
  // arrow
  collapsable
    ? h("span", { class: ["arrow", open ? "down" : "right"] }, text)
    : null,
];

export const renderChildren = (
  children: ResolvedSidebarHeaderItem[]
): VNode | null => {
  const route = useRoute();
  if (!children) return null;

  return h(
    "ul",
    { class: "sidebar-sub-headers" },
    children.map((child) => {
      const active = isActiveLink(route, child.link);

      return h("li", { class: "sidebar-sub-header" }, [
        renderItem(child, {
          class: {
            "sidebar-link": true,
            heading: true,
            active,
          },
        }),
        renderChildren(child.children),
      ]);
    })
  );
};
