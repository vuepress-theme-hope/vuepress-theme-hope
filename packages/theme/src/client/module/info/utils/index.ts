export * from "./define";
export * from "./editlink";
