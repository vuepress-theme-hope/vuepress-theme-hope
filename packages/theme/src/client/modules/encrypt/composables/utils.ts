import { type ComputedRef, computed } from "vue";

import { useThemeData } from "@theme-hope/composables/index";

import { type EncryptConfig } from "../../../../shared/index.js";

export const useEncryptData = (): ComputedRef<EncryptConfig> => {
  const themeData = useThemeData();

  return computed(() => themeData.value.encrypt || {});
};
