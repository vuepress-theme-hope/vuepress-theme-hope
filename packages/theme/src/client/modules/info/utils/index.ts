export * from "./editLink.js";
export * from "./typings.js";
