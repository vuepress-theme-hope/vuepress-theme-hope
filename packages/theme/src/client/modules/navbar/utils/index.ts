import {
  type AutoLinkOptions,
  type NavGroup,
  type NavbarItem,
} from "../../../../shared/index.js";

// resolved
export type ResolvedThemeNavbarItem =
  | NavbarItem
  | NavGroup<AutoLinkOptions | NavGroup<AutoLinkOptions>>;
