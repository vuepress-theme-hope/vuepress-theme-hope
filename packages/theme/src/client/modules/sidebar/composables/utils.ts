import { ensureEndingSlash } from "@vuepress/shared";

export const resolvePrefix = (prefix = "", path = ""): string =>
  path.startsWith("/") ? path : `${ensureEndingSlash(prefix)}${path}`;
