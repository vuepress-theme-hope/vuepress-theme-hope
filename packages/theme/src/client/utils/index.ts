export * from "./getAncestorLinks.js";
export * from "./resolveLinkInfo.js";
