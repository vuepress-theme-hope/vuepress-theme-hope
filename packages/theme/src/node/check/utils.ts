export const PLUGIN_CHECKLIST = [
  ["@vuepress/plugin-active-header-links", "activeHeaderLinks"],
  ["@vuepress/plugin-blog", "blog"],
  ["@vuepress/plugin-catalog", "catalog"],
  ["@vuepress/plugin-comment", "comment"],
  ["@vuepress/plugin-copy-code", "copyCode"],
  ["@vuepress/plugin-copyright", "copyright"],
  ["@vuepress/plugin-docsearch", "docsearch"],
  ["@vuepress/plugin-external-link-icon", "externalLinkIcon"],
  ["@vuepress/plugin-feed", "feed"],
  ["@vuepress/plugin-git", "git"],
  ["@vuepress/plugin-links-check", "linksCheck"],
  ["@vuepress/plugin-notice", "notice"],
  ["@vuepress/plugin-nprogress", "nprogress"],
  ["@vuepress/plugin-photo-swipe", "photoSwipe"],
  ["@vuepress/plugin-prismjs", "prismjs"],
  ["@vuepress/plugin-pwa", "pwa"],
  ["@vuepress/plugin-reading-time", "readingTime"],
  ["@vuepress/plugin-redirect", "redirect"],
  ["@vuepress/plugin-rtl", "", 'Set "rtl: true" in the needed theme locales.'],
  ["@vuepress/plugin-search", "search"],
  ["@vuepress/plugin-seo", "seo"],
  ["@vuepress/plugin-sitemap", "sitemap"],
  ["@vuepress/plugin-theme-data", "", "This plugin is called internally."],
  ["@vuepress/plugin-watermark", "watermark"],
  ["vuepress-plugin-components", "components"],
  ["vuepress-plugin-md-enhance", "mdEnhance"],
  ["vuepress-plugin-search-pro", "searchPro"],
];
