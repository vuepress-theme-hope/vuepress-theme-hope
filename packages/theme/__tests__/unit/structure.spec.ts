import { createBaseApp } from "@vuepress/core";
import { path } from "@vuepress/utils";
import { describe, expect, it } from "vitest";

import { emptyTheme } from "./__fixtures__/theme/empty.js";
import { getStructureInfo } from "../../src/node/prepare/index.js";

describe("should generate structure info correctly", async () => {
  const app = createBaseApp({
    bundler: {} as any,
    source: path.resolve(__dirname, "./__fixtures__/src"),
    theme: emptyTheme,
  });

  await app.init();

  it("Should generate structure info for whole dir", () => {
    expect(getStructureInfo(app.pages, "")).toMatchSnapshot();
  });

  it("Should resolve structure in dir", () => {
    expect(getStructureInfo(app.pages, "dir/")).toMatchSnapshot();
  });
});
