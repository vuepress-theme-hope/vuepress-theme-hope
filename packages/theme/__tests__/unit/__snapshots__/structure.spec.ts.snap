// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate structure info correctly > Should generate structure info for whole dir 1`] = `
[
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "dir-negative-order/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "dir-negative-order/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "dir-negative-order/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "dir-negative-order/file4.md",
        "type": "file",
      },
      {
        "filename": "README.md",
        "path": "dir-negative-order/README.md",
        "type": "file",
      },
    ],
    "dirname": "dir-negative-order",
    "path": "dir-negative-order",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "dir-no-order/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "dir-no-order/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "dir-no-order/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "dir-no-order/file4.md",
        "type": "file",
      },
      {
        "filename": "README.md",
        "path": "dir-no-order/README.md",
        "type": "file",
      },
    ],
    "dirname": "dir-no-order",
    "path": "dir-no-order",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "dir-positive-order/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "dir-positive-order/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "dir-positive-order/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "dir-positive-order/file4.md",
        "type": "file",
      },
      {
        "filename": "README.md",
        "path": "dir-positive-order/README.md",
        "type": "file",
      },
    ],
    "dirname": "dir-positive-order",
    "path": "dir-positive-order",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "dir/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "dir/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "dir/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "dir/file4.md",
        "type": "file",
      },
      {
        "filename": "permalink.md",
        "path": "dir/permalink.md",
        "type": "file",
      },
      {
        "filename": "permalinkpattern.md",
        "path": "dir/permalinkpattern.md",
        "type": "file",
      },
      {
        "filename": "README.md",
        "path": "dir/README.md",
        "type": "file",
      },
    ],
    "dirname": "dir",
    "path": "dir",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "component.md",
        "path": "excerpt/component.md",
        "type": "file",
      },
      {
        "filename": "cut.md",
        "path": "excerpt/cut.md",
        "type": "file",
      },
      {
        "filename": "demo.md",
        "path": "excerpt/demo.md",
        "type": "file",
      },
      {
        "filename": "example.md",
        "path": "excerpt/example.md",
        "type": "file",
      },
      {
        "filename": "link.md",
        "path": "excerpt/link.md",
        "type": "file",
      },
    ],
    "dirname": "excerpt",
    "path": "excerpt",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "link/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "link/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "link/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "link/file4.md",
        "type": "file",
      },
      {
        "filename": "README.md",
        "path": "link/README.md",
        "type": "file",
      },
    ],
    "dirname": "link",
    "path": "link",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "nested-dir/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "nested-dir/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "nested-dir/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "nested-dir/file4.md",
        "type": "file",
      },
      {
        "children": [
          {
            "filename": "file1.md",
            "path": "nested-dir/nested-dir1/file1.md",
            "type": "file",
          },
          {
            "filename": "file2.md",
            "path": "nested-dir/nested-dir1/file2.md",
            "type": "file",
          },
          {
            "filename": "file3.md",
            "path": "nested-dir/nested-dir1/file3.md",
            "type": "file",
          },
          {
            "filename": "file4.md",
            "path": "nested-dir/nested-dir1/file4.md",
            "type": "file",
          },
          {
            "children": [
              {
                "filename": "file1.md",
                "path": "nested-dir/nested-dir1/nested-dir2/file1.md",
                "type": "file",
              },
              {
                "filename": "file2.md",
                "path": "nested-dir/nested-dir1/nested-dir2/file2.md",
                "type": "file",
              },
              {
                "filename": "file3.md",
                "path": "nested-dir/nested-dir1/nested-dir2/file3.md",
                "type": "file",
              },
              {
                "filename": "file4.md",
                "path": "nested-dir/nested-dir1/nested-dir2/file4.md",
                "type": "file",
              },
              {
                "filename": "README.md",
                "path": "nested-dir/nested-dir1/nested-dir2/README.md",
                "type": "file",
              },
            ],
            "dirname": "nested-dir2",
            "path": "nested-dir/nested-dir1/nested-dir2",
            "type": "dir",
          },
          {
            "children": [
              {
                "filename": "file1.md",
                "path": "nested-dir/nested-dir1/nested-no-readme2/file1.md",
                "type": "file",
              },
              {
                "filename": "file2.md",
                "path": "nested-dir/nested-dir1/nested-no-readme2/file2.md",
                "type": "file",
              },
              {
                "filename": "file3.md",
                "path": "nested-dir/nested-dir1/nested-no-readme2/file3.md",
                "type": "file",
              },
              {
                "filename": "file4.md",
                "path": "nested-dir/nested-dir1/nested-no-readme2/file4.md",
                "type": "file",
              },
            ],
            "dirname": "nested-no-readme2",
            "path": "nested-dir/nested-dir1/nested-no-readme2",
            "type": "dir",
          },
          {
            "filename": "README.md",
            "path": "nested-dir/nested-dir1/README.md",
            "type": "file",
          },
        ],
        "dirname": "nested-dir1",
        "path": "nested-dir/nested-dir1",
        "type": "dir",
      },
      {
        "children": [
          {
            "filename": "file1.md",
            "path": "nested-dir/nested-no-readme1/file1.md",
            "type": "file",
          },
          {
            "filename": "file2.md",
            "path": "nested-dir/nested-no-readme1/file2.md",
            "type": "file",
          },
          {
            "filename": "file3.md",
            "path": "nested-dir/nested-no-readme1/file3.md",
            "type": "file",
          },
          {
            "filename": "file4.md",
            "path": "nested-dir/nested-no-readme1/file4.md",
            "type": "file",
          },
          {
            "children": [
              {
                "filename": "file1.md",
                "path": "nested-dir/nested-no-readme1/nested-no-readme2/file1.md",
                "type": "file",
              },
              {
                "filename": "file2.md",
                "path": "nested-dir/nested-no-readme1/nested-no-readme2/file2.md",
                "type": "file",
              },
              {
                "filename": "file3.md",
                "path": "nested-dir/nested-no-readme1/nested-no-readme2/file3.md",
                "type": "file",
              },
              {
                "filename": "file4.md",
                "path": "nested-dir/nested-no-readme1/nested-no-readme2/file4.md",
                "type": "file",
              },
            ],
            "dirname": "nested-no-readme2",
            "path": "nested-dir/nested-no-readme1/nested-no-readme2",
            "type": "dir",
          },
        ],
        "dirname": "nested-no-readme1",
        "path": "nested-dir/nested-no-readme1",
        "type": "dir",
      },
      {
        "filename": "README.md",
        "path": "nested-dir/README.md",
        "type": "file",
      },
    ],
    "dirname": "nested-dir",
    "path": "nested-dir",
    "type": "dir",
  },
  {
    "children": [
      {
        "filename": "file1.md",
        "path": "options/file1.md",
        "type": "file",
      },
      {
        "filename": "file2.md",
        "path": "options/file2.md",
        "type": "file",
      },
      {
        "filename": "file3.md",
        "path": "options/file3.md",
        "type": "file",
      },
      {
        "filename": "file4.md",
        "path": "options/file4.md",
        "type": "file",
      },
      {
        "filename": "README.md",
        "path": "options/README.md",
        "type": "file",
      },
    ],
    "dirname": "options",
    "path": "options",
    "type": "dir",
  },
  {
    "filename": "README.md",
    "path": "README.md",
    "type": "file",
  },
]
`;

exports[`should generate structure info correctly > Should resolve structure in dir 1`] = `
[
  {
    "filename": "file1.md",
    "path": "file1.md",
    "type": "file",
  },
  {
    "filename": "file2.md",
    "path": "file2.md",
    "type": "file",
  },
  {
    "filename": "file3.md",
    "path": "file3.md",
    "type": "file",
  },
  {
    "filename": "file4.md",
    "path": "file4.md",
    "type": "file",
  },
  {
    "filename": "permalink.md",
    "path": "permalink.md",
    "type": "file",
  },
  {
    "filename": "permalinkpattern.md",
    "path": "permalinkpattern.md",
    "type": "file",
  },
  {
    "filename": "README.md",
    "path": "README.md",
    "type": "file",
  },
]
`;
